### PR TITLE
✨ JAVA-3297 Create Scan Project If Not Exists

### PR DESCRIPTION
**Pull request by gilday**
Looks up an existing Scan project with the configured name, or creates a new Java project if no such project exists.

Because this code will eventually migrate to the SDK, I have been creating scan types (e.g. `Scan`, `Project`, `ScanSummary`) that include all known fields vs just the fields this project needs. I diverged from this approach when it comes to the paged results. The paging metadata returned by the Scan API doesn't feel right, and I'd like a chance to work with the Scan team to improve it before we further cement it into the Contrast Java SDK. Therefore, the paged result structure is a package-private type in this PR.

It seems the `ScanSummary` type now returns timestamps with time zone information, so we may not need to the `LocalDateTime` GSON adapter added previously. I am working with the Scan team to confirm.

### CODE REVIEW

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/Project.java [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+  /**
+   * @return count of note severity vulnerabilities detected in the last scan to complete
+   *     successfully
+   */
+  public int getNote() {
+    return note;
+  }
+
+  /** @return the time at which the last successfully completed scan finished */
```
Comment (stiemannkj1) [inline src/main/java/com/contrastsecurity/maven/plugin/sdkx/Project.java]: I think this method and `getLastScanId` can possibly return `null`, so should the docs call out that they might return `null` if there’s no previous successful scan?

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/Project.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
+package com.contrastsecurity.maven.plugin.sdkx;
+
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+
+/** Value object that describes a Contrast Scan project */
+public final class Project {
```
Comment (stiemannkj1) [inline src/main/java/com/contrastsecurity/maven/plugin/sdkx/Project.java:10]: I think we should consider naming this `ScanProject` since that will distinguish it from a local Maven project. I guess we don’t have anything called `Project` already in the Team Server API, but it also helps distinguish it from an Agent Application as Application and Project are somewhat generic terms. Maybe not quite interchangeable, but I could see users confusing them.

File: src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java [hunk 3]
```diff
@@ -41,15 +45,14 @@
...
 
   /**
    * Contrast Scan project unique ID to which the plugin runs new Scans. This will be replaced
    * imminently with a project name
    */
-  // TODO[JAVA-3297] replace this with "project"
-  @Parameter(required = true)
-  private String projectId;
+  @Parameter(property = "project", defaultValue = "${project.name}")
+  private String projectName;
```
Comment (stiemannkj1) [inline src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java:55]: Similar to `Project`, maybe we should consider prefixing these types of values with `scan` (`scanProjectName`). Maybe it’s overkill (since the docs make it unambiguous), but I like the idea that you can look at the Maven XML elements and know that we’re talking about the Scan project and not the Maven project.

File: src/test/java/com/contrastsecurity/maven/plugin/it/stub/FakeContrastAPI.java [hunk 4]
```diff
@@ -165,6 +232,13 @@ private static void downloadAgent(final HttpExchange exchange) {
     exchange.close();
```
Comment (stiemannkj1) [inline src/test/java/com/contrastsecurity/maven/plugin/it/stub/FakeContrastAPI.java]: Minor, but we could create another lambda wrapper factory for `HttpHandler` like `cleanupHandler`:

```
private static HttpHandler cleanupHandler(final HttpHandler httpHandler) {
    return exchange -> {
        try {
            discardRequest(exchange);
            httpHandler.handle(exchange);
        } finally {
            exchange.close();
        }
    }
}
```

That might encapsulate the boilerplate a bit.

File: src/test/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactoryTest.java [hunk 5]
```diff
@@ -19,14 +21,59 @@ void parse_iso8601_strings_to_local_date_time() {
...
     assertThat(obj.time).isEqualTo(expected);
   }
 
+  @Test
+  void parse_null_to_local_date_time() {
+    // GIVEN a new GSON instance that should handle java.time.LocalDateTime
+    final Gson gson = GsonFactory.create();
+
+    // WHEN deserialize a type containing a LocalDateTime that is null
+    final LocalDateTimeObj obj = gson.fromJson("{\"time\": null}", LocalDateTimeObj.class);
```
Comment (stiemannkj1) [inline src/test/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactoryTest.java]: Good catch with this. Didn’t realize that the handlers always need the deal with potential null values.

File: src/test/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactoryTest.java [hunk 6]
```diff
@@ -19,14 +21,59 @@ void parse_iso8601_strings_to_local_date_time() {
...
-  private static final class Obj {
+  private static final class LocalDateTimeObj {
     LocalDateTime time;
   }
+
+  /** GSON cannot work with local classes */
+  private static final class ZonedDateTimeObj {
+    ZonedDateTime time;
+  }
 }
```
Comment (stiemannkj1) [inline src/test/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactoryTest.java:85]: I’m not sure that we use it anywhere, but should we also test the writing of `LocalDateTime` and `ZonedDateTime` values? You did implement that behavior, so maybe it should be tested (along with writing `null` as well).

**Review by BrianPhillips2020 — APPROVED** [review]

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java [hunk 7]
```diff
@@ -38,6 +40,46 @@ public DefaultContrastScanSDK(final ContrastSDK contrast, final String restURL)
...
+  }
+
+  @Override
+  public Project findProjectByName(final String organizationId, final String projectName)
+      throws IOException, UnauthorizedException {
+    // requests made with ContrastSDK.makeRequest must have their path prepended with "/"
+    final String uri =
+        String.join("/", "", "sast", "organizations", organizationId, "projects")
+            + "?unique=true&name="
+            + URLEncoder.encode(projectName, "US-ASCII");
```
Comment (stiemannkj1) [inline src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java]: Can we use `StandardCharsets.US_ASCII` here? Also why not `UTF-8`?

File: src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java [hunk 8]
```diff
@@ -154,13 +163,70 @@ public void execute() throws MojoExecutionException, MojoFailureException {
     }
   }
 
+  /**
+   * Finds a Scan project with the project name from the plugin configuration, or creates such a
+   * "Java" project if one does not exist.
+   *
+   * <p>Note: the Scan API does not expose an endpoint for doing this atomically, so it is possible
+   * that another process creates the project after having determined it to not-exist but before
+   * attempting to create it.
```
Comment (stiemannkj1) [inline src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java:172]: Is this something the Umbrella team needs to address? Seems like it could cause confusion and I don’t believe we can solve it on our end. Really they need to implement an atomic `findOrCreateProject` endpoint that handles this case I guess.

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java [hunk 9]
```diff
@@ -38,6 +40,46 @@ public DefaultContrastScanSDK(final ContrastSDK contrast, final String restURL)
...
+    }
+  }
+
+  @Override
+  public Project findProjectByName(final String organizationId, final String projectName)
+      throws IOException, UnauthorizedException {
+    // requests made with ContrastSDK.makeRequest must have their path prepended with "/"
+    final String uri =
+        String.join("/", "", "sast", "organizations", organizationId, "projects")
+            + "?unique=true&name="
```
Comment (stiemannkj1) [inline src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java:62]: `unique=true` implies that you can have multiple projects with the same name. Seems like a lot of potential for confusion there. Maybe that’s another area where the umbrella API can improve? Or maybe I’m misunderstanding.

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java (hunk 7 above)
Comment (stiemannkj1) [reply src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java]: Actually, should we be using the Maven source encoding value: https://stackoverflow.com/questions/3017695/how-to-configure-encoding-in-maven?

I guess that’s the encoding for Java source though. Not sure it would apply to the project name in Maven XML.

**Review by stiemannkj1 — APPROVED** [review]
Looks great to me. Added some minor comments throughout. Only thing that could potentially be a bug would be the url encoding thing that I’ve called out within the PR.

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/Project.java (hunk 2 above)
Comment (gilday) [reply src/main/java/com/contrastsecurity/maven/plugin/sdkx/Project.java:10]: I'd like to be consistent with the names we use in everywhere else, even if those names are imperfect.

> since that will distinguish it from a local Maven project

I ran into this in the `ContrastScanMojo`. You'll see that I named the `MavenProject` `mavenProject` to distinguish the two. Recall that this type will migrate to the SDK which is unaware of Maven's domain models.

File: src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java (hunk 3 above)
Comment (gilday) [reply src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java:55]: I see what you mean, but again I'd like to be consistent with the names as they'll appear elsewhere. I think the context will help, and I'd like to think that most users will accept the default (which is the Maven project name)

File: src/test/java/com/contrastsecurity/maven/plugin/it/stub/FakeContrastAPI.java (hunk 4 above)
Comment (gilday) [reply src/test/java/com/contrastsecurity/maven/plugin/it/stub/FakeContrastAPI.java]: There are three methods that repeat this behavior, but there are six call sites that would need to be decorated. Which do you believe is the lesser of two evils?

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java (hunk 9 above)
Comment (gilday) [reply src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java:62]: Yes, that was an issue, and the team resolved it. This `unique` parameter is another change that team made for us. the name is imperfect, but it means "perform an exact match on the project name instead a 'contains' match". Without this parameter, if you search for "t", you won't get a project named "t", you'll get every project that starts with "t". The `unique` parameter is the only way they give us to look-up a project by name.

I'll send you more context in Slack

File: src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java (hunk 8 above)
Comment (gilday) [reply src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java:172]: That's right. I will raise it as a concern

File: src/test/java/com/contrastsecurity/maven/plugin/it/stub/FakeContrastAPI.java (hunk 4 above)
Comment (stiemannkj1) [reply src/test/java/com/contrastsecurity/maven/plugin/it/stub/FakeContrastAPI.java]: I like having cleanup encapsulated as its own thing personally (I think it makes it clearer that we need to perform these steps for each request/response), but if you think that’s overkill, I’m fine with leaving as-is as it’s not much duplicated code like you said.

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactory.java [hunk 10]
```diff
@@ -7,20 +7,37 @@
...
+    }
+
+    @Override
+    public ZonedDateTime read(final JsonReader reader) throws IOException {
+      final String iso8601 = reader.nextString();
+      return ZonedDateTime.parse(iso8601, DateTimeFormatter.ISO_ZONED_DATE_TIME);
+    }
+  }
+
   private static final class LocalDateTimeTypeAdapter extends TypeAdapter<LocalDateTime> {
```
Comment (stiemannkj1) [inline src/main/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactory.java]: I think we’ve determined that this `LocalDateTime` stuff necessary anymore right? Up to you if you want to remove it now or wait until later. I know you just asked about this in slack, so just putting a comment here as a reminder.

File: src/test/java/com/contrastsecurity/maven/plugin/it/stub/FakeContrastAPI.java (hunk 4 above)
Comment (gilday) [reply src/test/java/com/contrastsecurity/maven/plugin/it/stub/FakeContrastAPI.java]: OK. I'll make this change and try to reduce any boilerplate as best I can

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java (hunk 7 above)
Comment (gilday) [reply src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java]: Right, the Maven source encoding is the encoding for the source files.

If I've read the HTTP specification correctly, then URIs in the request line use US-ASCII

We cannot use StandardCharsets.US_ASCII because that method overload was introduced in Java 10 https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URLEncoder.html#encode(java.lang.String,java.nio.charset.Charset) 

I guess I can use

```java
URLEncoder.encode(projectName, StandardCharsets.US_ASCII.name());
```

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactory.java (hunk 10 above)
Comment (gilday) [reply src/main/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactory.java]: Yes, the Scan API team informed me that they fixed the API. I'll get rid of the `LocalDateTime` stuff. I think I'm wrong to use `ZonedDateTime` - I should be using `OffsetDateTime` - so I'll make that change too.

File: src/test/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactoryTest.java (hunk 6 above)
Comment (gilday) [reply src/test/java/com/contrastsecurity/maven/plugin/sdkx/GsonFactoryTest.java:85]: I was thinking of changing the `write` code to throw `UnsupportedOperationException` so I didn't have to worry about it, but since this code is destined for the SDK where we very well may have a use case for serializing these types, I'll just add those tests instead.

File: src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java (hunk 7 above)
Comment (gilday) [reply src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java]: I've added a test for this

Comment (stiemannkj1) [reply src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java]: 👍 can you add a test case with a project name that includes unicode chars? I think that test case would fail since we’d translate all UTF-8 chars to `?` (or `%3F`). 

The JavaDoc for `URLEncoder.encode` recommends `UTF-8` for this reason I think: https://docs.oracle.com/javase/7/docs/api/java/net/URLEncoder.html#encode(java.lang.String,%20java.lang.String)

Comment (gilday) [reply src/main/java/com/contrastsecurity/maven/plugin/sdkx/DefaultContrastScanSDK.java]: OK we got to the bottom of this. It should definitely be UTF-8. Test added
